File: src/go/svd/extensions/service_provider.go
package extensions

import (
	"github.com/WMS-Ceataec/standardised-vessel-dataset/src/go/svd/exporters"
	csvExporter "github.com/WMS-Ceataec/standardised-vessel-dataset/src/go/svd/exporters/formats/csv"
	jsonExporter "github.com/WMS-Ceataec/standardised-vessel-dataset/src/go/svd/exporters/formats/json"
	xmlExporter "github.com/WMS-Ceataec/standardised-vessel-dataset/src/go/svd/exporters/formats/xml"
	"github.com/WMS-Ceataec/standardised-vessel-dataset/src/go/svd/validators"
)

// ServiceProvider provides a simple service container for SVD components
type ServiceProvider struct {
	validator    validators.SVDValidator
	jsonExporter exporters.SVDJSONExporter
	csvExporter  exporters.SVDCSVExporter
	xmlExporter  exporters.SVDXMLExporter
}

// NewServiceProvider creates a new service provider with default implementations
func NewServiceProvider() *ServiceProvider {
	validator := validators.NewStandardisedVesselDatasetValidator()

	return &ServiceProvider{
		validator:    validator,
		jsonExporter: jsonExporter.NewJSONExporter(validator),
		csvExporter:  csvExporter.NewCSVExporter(validator),
		xmlExporter:  xmlExporter.NewXMLExporter(validator),
	}
}

// NewServiceProviderWithCustomValidator creates a new service provider with custom validator
func NewServiceProviderWithCustomValidator(validator validators.SVDValidator) *ServiceProvider {
	return &ServiceProvider{
		validator:    validator,
		jsonExporter: jsonExporter.NewJSONExporter(validator),
		csvExporter:  csvExporter.NewCSVExporter(validator),
		xmlExporter:  xmlExporter.NewXMLExporter(validator),
	}
}

// GetValidator returns the SVD validator
func (s *ServiceProvider) GetValidator() validators.SVDValidator {
	return s.validator
}

// GetJSONExporter returns the JSON exporter
func (s *ServiceProvider) GetJSONExporter() exporters.SVDJSONExporter {
	return s.jsonExporter
}

// GetCSVExporter returns the CSV exporter
func (s *ServiceProvider) GetCSVExporter() exporters.SVDCSVExporter {
	return s.csvExporter
}

// GetXMLExporter returns the XML exporter
func (s *ServiceProvider) GetXMLExporter() exporters.SVDXMLExporter {
	return s.xmlExporter
}

// SetValidator sets a custom validator
func (s *ServiceProvider) SetValidator(validator validators.SVDValidator) {
	s.validator = validator
	// Update exporters with new validator
	s.jsonExporter = jsonExporter.NewJSONExporter(validator)
	s.csvExporter = csvExporter.NewCSVExporter(validator)
	s.xmlExporter = xmlExporter.NewXMLExporter(validator)
}

// SetJSONExporter sets a custom JSON exporter
func (s *ServiceProvider) SetJSONExporter(exporter exporters.SVDJSONExporter) {
	s.jsonExporter = exporter
}

// SetCSVExporter sets a custom CSV exporter
func (s *ServiceProvider) SetCSVExporter(exporter exporters.SVDCSVExporter) {
	s.csvExporter = exporter
}

// SetXMLExporter sets a custom XML exporter
func (s *ServiceProvider) SetXMLExporter(exporter exporters.SVDXMLExporter) {
	s.xmlExporter = exporter
}

File: src/elixir/svd/lib/svd/extensions/service_provider.ex
defmodule SVD.Extensions.ServiceProvider do
  @moduledoc false

  alias SVD.Exporters.Formats.{CSVExporter, JSONExporter, XMLExporter}
  alias SVD.Validators.StandardisedVesselDatasetValidator

  @enforce_keys [:validator, :json_exporter, :csv_exporter, :xml_exporter]
  defstruct [:validator, :json_exporter, :csv_exporter, :xml_exporter]

  @type t :: %__MODULE__{}

  @spec new() :: t()
  def new do
    new_with_custom_validator(StandardisedVesselDatasetValidator)
  end

  @spec new_with_custom_validator(module()) :: t()
  def new_with_custom_validator(validator_module) when is_atom(validator_module) do
    %__MODULE__{
      validator: validator_module,
      json_exporter: JSONExporter.new(validator_module),
      csv_exporter: CSVExporter.new(validator_module),
      xml_exporter: XMLExporter.new(validator_module)
    }
  end

  @spec get_validator(t()) :: module()
  def get_validator(%__MODULE__{validator: validator}), do: validator

  @spec get_json_exporter(t()) :: JSONExporter.t()
  def get_json_exporter(%__MODULE__{json_exporter: exporter}), do: exporter

  @spec get_csv_exporter(t()) :: CSVExporter.t()
  def get_csv_exporter(%__MODULE__{csv_exporter: exporter}), do: exporter

  @spec get_xml_exporter(t()) :: XMLExporter.t()
  def get_xml_exporter(%__MODULE__{xml_exporter: exporter}), do: exporter

  @spec set_validator(t(), module()) :: t()
  def set_validator(%__MODULE__{} = provider, validator_module) when is_atom(validator_module) do
    %{
      provider
      | validator: validator_module,
        json_exporter: JSONExporter.new(validator_module),
        csv_exporter: CSVExporter.new(validator_module),
        xml_exporter: XMLExporter.new(validator_module)
    }
  end

  @spec set_json_exporter(t(), JSONExporter.t()) :: t()
  def set_json_exporter(%__MODULE__{} = provider, exporter),
    do: %{provider | json_exporter: exporter}

  @spec set_csv_exporter(t(), CSVExporter.t()) :: t()
  def set_csv_exporter(%__MODULE__{} = provider, exporter),
    do: %{provider | csv_exporter: exporter}

  @spec set_xml_exporter(t(), XMLExporter.t()) :: t()
  def set_xml_exporter(%__MODULE__{} = provider, exporter),
    do: %{provider | xml_exporter: exporter}
end
